atlasdb-commons/src/test/java/com/palantir/common/proxy/InterruptibleProxyTest.java

```diff
@@ -0,0 +1,118 @@
+        final CyclicBarrier barrier = new CyclicBarrier(2);
+        final Thread callingThread = Thread.currentThread();
+        final AtomicBoolean gotInterrupted = new AtomicBoolean(false);
+        List<String> strings = new ArrayList<String>() {
+            private static final long serialVersionUID = 1L;
+
+            @Override
+            public String get(int arg0) {
+                if (arg0 == 1) {
+                    // In order for the get(1) test below to reliably check
+                    // interrupt we can't actually finish (otherwise our
+                    // completion is racing main thread checking the future and
+                    // if we win it gets the value instead of Future.get()
+                    // checking interrupt).
+                    try {
+                        Thread.sleep(Long.MAX_VALUE);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(e);
+                    }
+                }
+                return super.get(arg0);
+            }
+
+            @Override
+            public String remove(int arg0) {
+                assertNotEquals(Thread.currentThread(), callingThread);
+                while(true) {
+                    try {
+                        Thread.sleep(100);
+                        callingThread.interrupt();
+                    } catch (InterruptedException e) {
+                        gotInterrupted.set(true);
+                        try {
+                            barrier.await();
+                        } catch (InterruptedException e1) {
+                            Thread.currentThread().interrupt();
+                            throw new RuntimeException(e1);
+                        } catch (BrokenBarrierException e1) {
+                            throw new RuntimeException(e1);
+                        }
+                        throw new PalantirRuntimeException(e);
+                    }
+                }
+            }
+        };
+        strings.add("Foo");
+
+        @SuppressWarnings("unchecked")
+        List<String> proxy =
+                InterruptibleProxy.newProxyInstance(List.class, strings, CancelDelegate.CANCEL);
+        assertEquals("Foo", proxy.get(0));
+        try {
+            Thread.currentThread().interrupt();
+            assertEquals("Foo", proxy.get(1));
+            fail("Should be interrupted");
+        } catch (PalantirInterruptedException e) {
+            // Should get here
+        }
+
+        try {
+            assertEquals("Foo", proxy.get(1000));
+            fail("Should throw exception");
+        } catch (IndexOutOfBoundsException e) {
+            // Should get here
+        }
+
+        try {
+            proxy.remove(0);
+            fail("Should be interrupted");
+        } catch (PalantirInterruptedException e) {
+            // Should get here
+        }
+
+        barrier.await();
```

Comment:
It's unclear to me what the barrier is doing for us, since `gotInterrupted` will get set regardless of whether the proxy thread gets stuck waiting on the barrier.